server/src/main/java/org/elasticsearch/action/support/RefCountingListener.java

```diff
@@ -187,10 +187,12 @@ public void onResponse(Response response) {
                     if (acquiredConsumer == null) {
                         assert false : "already closed";
                     } else {
-                        acquiredConsumer.accept(response);
+                        try {
+                            acquiredConsumer.accept(response);
+                        } catch (Exception e) {
+                            addException(e);
+                        }
                     }
-                } catch (Exception e) {
-                    addException(e);
```

Comment:
The bug was that this `catch` block executes after releasing the `ref` to which the `try` refers. We need to call `addException(e)` before releasing the last `ref`.